File: clients/python/src/zigbase/_jwt.py
"""JWT payload decoding, without signature verification.

Port of clients/typescript/src/jwt.ts. The client never verifies a token's
signature (only the server can, holding the secret); these helpers only read
claims to drive client-side behavior like pre-emptive refresh.
"""

from __future__ import annotations

import base64
import json
import time
from typing import Any


def decode_jwt_payload(token: str) -> dict[str, Any] | None:
    """Decode the payload segment of a JWT, or `None` for malformed input.

    Returns `None` for anything that isn't a string with exactly three
    dot-separated segments, an empty payload segment, invalid base64url, or
    a payload that doesn't decode to a JSON object.
    """
    parts = token.split(".")
    if len(parts) != 3 or not parts[1]:
        return None

    try:
        segment = parts[1]
        padded = segment + "=" * (-len(segment) % 4)
        decoded_bytes = base64.urlsafe_b64decode(padded)
        payload = json.loads(decoded_bytes.decode("utf-8"))
    except Exception:
        return None

    return payload if isinstance(payload, dict) else None


def is_token_expired(token: str, leeway_seconds: int = 0) -> bool:
    """Return `True` when `token` is expired (or has no readable `exp` claim).

    `leeway_seconds` is subtracted from `exp` before comparing to the
    current time, so a token can be treated as expired slightly before its
    real expiry (e.g. to account for clock skew or in-flight request
    latency).
    """
    payload = decode_jwt_payload(token)
    exp = payload.get("exp") if payload is not None else None
    if not isinstance(exp, (int, float)) or isinstance(exp, bool):
        return True

    now = int(time.time())
    return exp - leeway_seconds <= now


__all__ = ["decode_jwt_payload", "is_token_expired"]

File: clients/kotlin/src/main/kotlin/io/github/valthon/zigbase/jwt/Jwt.kt
package io.github.valthon.zigbase.jwt

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.Base64

/**
 * Decodes the payload segment of a JWT, without verifying its signature.
 *
 * Port of `decodeJwtPayload` in `clients/typescript/src/jwt.ts`. The client
 * never verifies a token's signature (only the server can, holding the
 * secret); this only reads claims to drive client-side behavior like
 * pre-emptive refresh.
 *
 * Returns `null` for any malformed input: a token that doesn't have exactly
 * three dot-separated segments, an empty payload segment, invalid
 * base64url, a payload segment that isn't valid UTF-8, or a payload that
 * doesn't decode to a JSON object.
 */
fun decodeJwtPayload(token: String): JsonObject? {
    val parts = token.split(".")
    if (parts.size != 3 || parts[1].isEmpty()) return null

    return try {
        val segment = parts[1]
        val padded = segment + "=".repeat((4 - segment.length % 4) % 4)
        val bytes = Base64.getUrlDecoder().decode(padded)

        val decoder =
            StandardCharsets.UTF_8
                .newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
        val text = decoder.decode(ByteBuffer.wrap(bytes)).toString()

        val element = Json.parseToJsonElement(text)
        element as? JsonObject
    } catch (e: Exception) {
        null
    }
}

/**
 * Returns `true` when [token] is expired (or has no readable numeric `exp`
 * claim).
 *
 * Port of `isTokenExpired` in `clients/typescript/src/jwt.ts`. [leewaySeconds]
 * is subtracted from `exp` before comparing to the current time, so a token
 * can be treated as expired slightly before its real expiry (e.g. to
 * account for clock skew or in-flight request latency). [nowEpochSeconds]
 * is injectable so callers (notably tests) can pin "now" instead of racing
 * the system clock; it defaults to the wall clock.
 */
fun isTokenExpired(
    token: String,
    leewaySeconds: Long = 0,
    nowEpochSeconds: () -> Long = { System.currentTimeMillis() / 1000 },
): Boolean {
    val payload = decodeJwtPayload(token) ?: return true
    val exp = payload["exp"] as? JsonPrimitive ?: return true
    // A JSON boolean is a JsonPrimitive too; content.toDoubleOrNull() rejects
    // "true"/"false" as a non-numeric claim, so booleans fall through here.
    if (exp.isString) return true
    val expSeconds = exp.content.toDoubleOrNull() ?: return true

    return expSeconds - leewaySeconds <= nowEpochSeconds()
}
